[SPARK-46179][SQL] Add CrossDbmsQueryTestSuites, which runs other DBMS against golden files, starting with Postgres

andylam-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Create `CrossDbmsQueryTestSuite`, which extends `SQLQueryTestHelper` and `DockerIntegrationSuite`. `CrossDbmsQueryTestSuite` is a trait class that allows testing golden files against other DBMS.

`PostgreSQLQueryTestSuite` is an implementation of `CrossDbmsQueryTestSuite`. For starters, sql files in the subquery sql-tests are automatically opted into this test. In this PR, all files except for `exists-having.sql` are ignored, otherwise this PR would have 10K+ line changes (I would like to do that in the next PR, if possible). I had to change the syntax for view creation in `exists-having.sql` slightly, and this is reflected in the `analyzer-results` file, but crucially, the query output (in the `results` file) are not changed.

Note that this will not be applicable to many of the current sql tests we have due to:
- Incompatible SQL syntax between spark sql and postgres.
- Incompatible data types.
- Difference in numerical precision with doubles.
- Missing functions in either system.
- Test files with specific configs, such as ANSI, count bug etc.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  3. If you fix a bug, you can clarify why it is a bug.
-->
For correctness checking of our SQLQueryTestSuites, we want to run SQLQueryTestSuites with Postgres as a reference DBMS. This can be easily extensible to other DBMS. 

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
This is a test-related PR, does not affect system behaviors. 

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No.

andylam-db committed:
Initial commit

andylam-db committed:
Add CrossDbmsQueryTestSuite

andylam-db committed:
Small changes for niceness

andylam-db committed:
Save point

andylam-db committed:
Add sqllogictest-select1.sql in postgres-crosstest input file

andylam-db committed:
Add header for class suite

andylam-db committed:
Small changes

andylam-db committed:
Trying to get the input files set up..

andylam-db committed:
Passing test!!

andylam-db committed:
Ignore all earlier tests first

andylam-db committed:
Small comment changes

andylam-db committed:
More comments

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:
```diff
@@ -928,14 +932,18 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
         output
     }
-    override def numSegments: Int = 3
+    override def numSegments: Int = if (schema.isDefined) { 3 } else { 2 }
```
please add a comment here describing the logic?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/JdbcConnection.scala:
```diff
@@ -0,0 +1,101 @@
+private[sql] trait JdbcConnection {
+  /**
+   * Runs the given query.
+   * @return A Seq[String] representing the output.
```
can you comment why this is a sequence of strings, rather than a single string? What are the components?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/SQLQueryTestRunner.scala:
```diff
@@ -0,0 +1,45 @@
+package org.apache.spark.sql.crossdbms
+
+/**
+ * Trait for classes that can run SQL queries for testing.
```
the method name implies this :) please extend to describe what this trait specifically does?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/SQLQueryTestRunner.scala:
```diff
@@ -0,0 +1,45 @@
+trait SQLQueryTestRunner {
+
+  /**
+   * Runs a given query and returns a Seq[String] that represents the query result output.
```
please explain why it's a sequence? what are the components?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+// scalastyle:on line.size.limit
+class CrossDbmsQueryTestSuite extends SQLQueryTestSuite with Logging {
+
+  // Note: the below two functions have to be functions instead of variables because the superclass
```
optional: you could make them `lazy val`s again to enforce that the result is only computed once. Up to you.

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+  // Note: the below two functions have to be functions instead of variables because the superclass
+  // runs the test first before the subclass variables can be instantiated.
+  private def crossDbmsToGenerateGoldenFiles: String = {
+    val userInputDbms = System.getenv("REF_DBMS")
```
can you put this magic constant into a separate `val` with a comment saying what it is?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+  // runs the test first before the subclass variables can be instantiated.
+  private def crossDbmsToGenerateGoldenFiles: String = {
+    val userInputDbms = System.getenv("REF_DBMS")
+    if (userInputDbms != null && userInputDbms.nonEmpty) {
```
you can do `Option(userInputDbms).map { dbms => ... }`

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+      CrossDbmsQueryTestSuite.DEFAULT_DBMS
+    }
+  }
+  private def customConnectionUrl: String = System.getenv("REF_DBMS_CONNECTION_URL")
```
can you put this magic constant into a separate val with a comment saying what it is?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+  private def customConnectionUrl: String = System.getenv("REF_DBMS_CONNECTION_URL")
+
+  override protected def runQueries(
+    queries: Seq[String],
```
please indent these function params by +4 total spaces

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+        // the already generated golden files.
+        if (regenerateGoldenFiles) {
+          if (runner.isEmpty) {
+            val connectionUrl = if (customConnectionUrl != null && customConnectionUrl.nonEmpty) {
```
you can do `Option(customConnectionUrl).map { url => ... }`

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+          }
+          val sparkDf = spark.sql(sql)
+          val output = runner.map(_.runQuery(sql)).get
+          // Use Spark analyzed plan to check if the query result is already semantically sorted
```
please update the comments to be complete sentences ending with punctuation (same everywhere else in this PR)

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+          }
+          result
+        } else {
+          handleExceptions(getNormalizedQueryExecutionResult(localSparkSession, sql))._2
```
please split this into separate statements to clarify?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+        // Don't care about the schema for this test. Only care about correctness.
+        schema = None,
+        output = normalizeTestResults(output.mkString("\n")))
+      if (testCase.isInstanceOf[CTETest]) {
```
please add a comment here describing why we care about CTE tests?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+    if (ignoreList.exists(t =>
+      testCase.name.toLowerCase(Locale.ROOT).contains(t.toLowerCase(Locale.ROOT)))) {
+      ignore(testCase.name) {
+        /* Do nothing */
```
maybe explicitly log that we are intentionally ignoring this filename?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+  }
+
+  override def createScalaTestCase(testCase: TestCase): Unit = {
+    if (ignoreList.exists(t =>
```
optional: you could just add this to the `testCase match` block below as the first case, to simplify the logic.

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+          }
+        case _ =>
+          ignore(s"Ignoring test cases that are not [[RegularTestCase]] for now") {
+            /* Do nothing */
```
Same, let's log it?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+  override lazy val listTestCases: Seq[TestCase] = {
+    listFilesRecursively(new File(inputFilePath)).flatMap { file =>
+      var resultFile = resultFileForInputFile(file)
+      // JDK-4511638 changes 'toString' result of Float/Double
```
is this code copied from somewhere? if so, please deduplicate it into one shared helper?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+    }.sortBy(_.name)
+  }
+
+  private def isSemanticallySorted(plan: LogicalPlan): Boolean = plan match {
```
please add a comment.

Are you sure this logic doesn't already exist somewhere? Please look in SQLQueryTestSuite?

It seems if we want to do this here, we probably want to do this for all the other golden file testing as well. (I agree with the idea, it makes test cases more deterministic.)

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+
+  // Ignore all tests for now due to likely incompatibility.
+  override def ignoreList: Set[String] = super.ignoreList ++ Set(
+    "postgreSQL",
```
please put the directories in a separate `Set[String]`, and sort the directories and .sql filenames alphabetically.

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,350 @@
+    case _ => plan.children.iterator.exists(isSemanticallySorted)
+  }
+
+  // Ignore all tests for now due to likely incompatibility.
```
How many tests do you think we will actually run with this mode later? Do you think an allowlist might make more sense than a denylist? For example, would just the golden files in the existing `postgres` directory maybe make sense for this?

andylam-db replied:
Unfortunately a lot of the golden files in the existing `postgres` directory have also been heavily changed to make it compatible to Spark. Disclaimer, I didn't run through all of them.

I think an ignoreList (although it's cumbersomely long) makes sense because it means that new SQL tests written would automatically be subject to this test. If the new SQL test isn't postgres compatible, then the developer would either modify the test (if it makes sense), or have to deliberately add it to the ignore list.

jchen5 replied:
It may be better to store this info with each test or output file, could be similar to how you had it in a previous version. When you run SQLQueryTestSuite to generate golden files, it could take an option to control whether it's run vs postgres or spark. When the test is first created the default can be postgres; if it isn't compatible the test author would just rerun with a flag. Afterwards it can reuse whatever mode was used previously by default.

But also I think the default should remain to test without postgres until we've simplified the process, e.g. provided some automation of the postgres installation.

dtenedor commented on sql/core/src/test/resources/sql-tests/inputs/postgres-crosstest/sqllogictest-select1.sql:
```diff
@@ -0,0 +1,90 @@
+  FROM t1
+ ORDER BY 1;
+
+SELECT a+b*2+c*3+d*4+e*5,
```
please fix formatting to e.g. add spaces before and after binary operators.

dtenedor reviewed (commented):
Thanks for adding this, it will be useful to increase test coverage!

andylam-db committed:
Address simple comments

andylam-db committed:
A little more small changes..

andylam-db commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/JdbcConnection.scala:
```diff
@@ -0,0 +1,101 @@
+          rows.append(row)
+        }
+      }
+      rows.map(_.mkString("\t")).toSeq
```
Tabs are used in `hiveResultString` as well, which is used in `SQLQuerySuite`. I followed it to reduce hassle, let me know if it's a big deal.

dtenedor replied:
OK, no problem

andylam-db commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/JdbcConnection.scala:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  def dropTable(tableName: String): Unit = {
+    val dropTableSql = s"DROP TABLE IF EXISTS $tableName"
```
Hmm, I wasn't too concerned about this/didn't think this through too much because this is only used in testing. Should I be?

andylam-db committed:
More small changes and removal of redundant code

andylam-db committed:
Generate golden files for SQLQueryTestSuite

andylam-db committed:
Revert "Generate golden files for SQLQueryTestSuite"

This reverts commit bef801184d51936a81d7b2c261cb71bafbdeed3e.

andylam-db committed:
Ignore "postgres-crosstest" in SQLQueryTestSuite

andylam-db committed:
Add comment to clarify why it is ignored

andylam-db committed:
Fix compilation failures

andylam-db committed:
Regenerate golden files for sqllogictest-select1.sql.out

andylam-db committed:
Tiny changes..

andylam-db committed:
Add postgresql back to ignorelist

andylam-db commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,330 @@
+          runner = runner.orElse(
+            Some(CrossDbmsQueryTestSuite.DBMS_TO_CONNECTION_MAPPING(
+              crossDbmsToGenerateGoldenFiles)(connectionUrl)))
+          val sparkDf = spark.sql(sql)
+          val output = runner.map(_.runQuery(sql)).get
+          // Use Spark analyzed plan to check if the query result is already semantically sorted.
+          if (isSemanticallySorted(sparkDf.queryExecution.analyzed)) {
```
This is problematic because Spark may error when Postgres doesn't, if, for example, the view/table was created in Postgres but not in Spark. I don't think checking whether a query has an `ORDER BY` clause is necessarily sufficient, however..

jchen5 replied:
If there is a difference in Spark vs Postgres, we should probably have golden file generation fail with an error (since otherwise, it would fail when you run the test anyway)

andylam-db committed:
Add exception handling in CrossDbmsQueryTestSuite

andylam-db committed:
Do refactoring so that we can add an additional input argument for CrossDbmsQueryTestSuite

andylam-db committed:
Generate with postgres

andylam-db committed:
Fix compilation error with ThriftServerQueryTestSuite

andylam-db committed:
Generate golden files with SQLQueryTestSuite

andylam-db committed:
Fixed a bug where tests weren't running against the golden file

andylam-db committed:
Use local spark session

andylam-db committed:
Small comment change

andylam-db committed:
Merge master

andylam-db committed:
Revert change in ThriftServerQueryTestSuite

andylam-db committed:
Remove DialectConverter..

andylam-db committed:
Don't run the tests if the cross dbms is not specified

agubichev commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:
```diff
@@ -98,20 +98,24 @@ trait SQLQueryTestHelper extends Logging {
+   * important to make non-sorted queries test cases more deterministic.
+   */
+  protected def isSemanticallySorted(plan: LogicalPlan): Boolean = plan match {
+    case _: Join | _: Aggregate | _: Generate | _: Sample | _: Distinct => false
```
how about window functions?

andylam-db replied:
This function was pulled out from below. Changing it might cause lots of golden files changes :O

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:
```diff
@@ -349,20 +349,16 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession with SQLHelper
     }
   }
 
```
Ping on this comment.

dtenedor replied:
Ping on this comment.

andylam-db replied:
I've been deliberating different ways of tackling this project, will make changes soon!

andylam-db replied:
@dtenedor Refactoring PR here https://github.com/apache/spark/pull/44405

andylam-db committed:
Remove sqllogictest

andylam-db committed:
Small niceness changes

andylam-db committed:
Do null -> NULL replacement, and put 2 tests in

andylam-db committed:
Add docker compose and bash scripts for easy postgres instance

andylam-db committed:
Trivial changes

andylam-db committed:
Trivial changes

andylam-db committed:
Fix typo

andylam-db committed:
Add header comments, and make CrossDbmsQuerySuite an abstract class

andylam-db committed:
Trivial changes

andylam-db committed:
Use prepared statements

andylam-db committed:
Remove sql file changes for now

andylam-db committed:
Header comment improvements

andylam-db committed:
Trivial changes

andylam-db committed:
Add custom postgres command

andylam-db committed:
Modify exists-having to be compatible with psql

andylam-db committed:
Merge branch 'master' into crossdbms

andylam-db committed:
Modify query slightly

andylam-db committed:
Add readme and add ONLY_IF for most sql files in subquery dir

andylam-db committed:
Rewrite exists-having

andylam-db committed:
Update comment

andylam-db committed:
Merge master

andylam-db commented:
@jchen5 @agubichev @dtenedor Pinging for re-reviews :-) There's are quite a lot of changes since last time (read PR description).
1. I changed `CrossDbmsQueryTestSuite` to be an abstract class, and then `PostgreSQLQueryTestSuite` to extend that.
2. For `PostgreSQLQueryTestSuite`, the default input directory is `sql-tests/subquery`, to reduce the scope. In this PR, all files except for `exists-having.sql` are ignored, otherwise this PR would have 10K+ line changes, because I have to change the view creation syntax in these files a bit to be compatible with Postgres.
3. I added a docker file and bash script for easy installation and start of a Postgres container.

andylam-db committed:
Ignore subquery-offset.sql

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/JdbcConnection.scala:
```diff
@@ -0,0 +1,124 @@
+
+  /**
+   * Loads data from the given Spark DataFrame into the table with the specified name.
+   * @param df        The Spark DataFrame containing the data to be loaded.
```
please fix indentation here?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,278 @@
+ * }}}
+ */
+
+class PostgreSQLQueryTestSuite extends CrossDbmsQueryTestSuite {
```
can you put this in a separate file, please?

dtenedor commented on sql/core/src/test/scala/org/apache/spark/sql/crossdbms/CrossDbmsQueryTestSuite.scala:
```diff
@@ -0,0 +1,278 @@
+ * which is eligible for the SQL file. These strings are defined in the companion object. For
+ * example, if you have a SQL file named `describe.sql`, and you want to indicate that Postgres is
+ * incompatible, add the following comment into the input file:
+ * --ONLY_IF spark
```
This could be simpler by just inferring whether to test against e.g. Postgres or not based on the presence of the corresponding golden file in its associated directory instead.

andylam-db replied:
I wanted to do this so that new sql test files are automatically opted into testing against Postgres. This "forces" the developer who adds the new test files to either generate golden files with Postgres, or actively ignore it.

andylam-db committed:
Create new file PostgreSQLQueryTestSuite.scala and fix indent

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-select.sql:
```diff
@@ -1,5 +1,6 @@
 -- A test suite for scalar subquery in SELECT clause
 
+--ONLY_IF spark
```
if most golden files are Spark only, shall we use opt-in approach and only run golden file with pgsql if it has a special marker?

cloud-fan commented:
one high-level question: how would the test system look like after we support more than one DBMS? Which DBMS we should pick to generate golden files? Or for each golden file, we only run it with one reference DBMS?

andylam-db commented:
> one high-level question: how would the test system look like after we support more than one DBMS? Which DBMS we should pick to generate golden files? Or for each golden file, we only run it with one reference DBMS?

@cloud-fan I'm thinking that each input SQL file will be subject to all DBMS, unless otherwise specified. This means that each input SQL file can produce more than one golden file (1 for Spark, then 1 for each DBMS).

cloud-fan commented:
> 1 for Spark, then 1 for each DBMS

Our goal is to guarantee Spark produces the same result as the reference DBMS, why do we want to have one golden file for each DBMS? What if they are different? Which one should Spark follow?

andylam-db commented:
> Our goal is to guarantee Spark produces the same result as the reference DBMS, why do we want to have one golden file for each DBMS? What if they are different? Which one should Spark follow?

That makes sense. I think all DBMS can share the same golden file, if we make the queries simple enough that they are compatible with all DBMS. I wish there was a way to only have one set of golden files, but the current golden files have the Spark schema and hence aren't able to be generated with the DBMS. I don't want to make too many changes to the golden file.

andylam-db committed:
Move from postgres-results -> crossdbms-results

cloud-fan commented:
> That makes sense. I think all DBMS can share the same golden file, if we make the queries simple enough that they are compatible with all DBMS.

Makes sense to me. Then shall we always let Spark generate golden files? What we need to do is to let other DBMS run the query and compare results with golden files (exclude the schema part). I think keeping a single golden file is important to verify the results, otherwise we need human to make sure two golden files are equal.

andylam-db commented:
> Then shall we always let Spark generate golden files? What we need to do is to let other DBMS run the query and compare results with golden files (exclude the schema part).

The reason why we wanted to use DBMS to generate golden files is because we didn't want the CI/CD/developer to have to install some version of postgres/other DBMS just to run tests.

andylam-db committed:
Fix silly paste

cloud-fan commented:
> because we didn't want the CI/CD/developer to have to install some version of postgres/other DBMS just to run tests.

Isn't it better to trust CI/CD to install the proper version of DBMS and run the tests, instead of humans? AFAIK we are already doing it to run tests for JDBC data sources. We already have postgres tests in https://github.com/apache/spark/blob/master/connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala

andylam-db committed:
Fix unintended removal of results file

andylam-db committed:
In middle of refactoring..

andylam-db committed:
Major refactoring in progress..

andylam-db committed:
Passing tests

andylam-db committed:
Add documnetation and delete previous files

andylam-db committed:
Move functions from PostgreSQLQueryTestSuite -> CrossDbmsQueryTestSuite

andylam-db committed:
Merge master..

andylam-db commented:
> We are already doing it to run tests for JDBC data sources

@cloud-fan I didn't know about those tests! I made the changes to add this as an integration test instead, as you suggested. The two tests are a bit disconnected:
1. It's maybe a bit weird that a test in `spark/connector` uses files in `spark/core`, but this can be improved in the future.
2. There's some code duplication to read golden files, but I think not too much that it is bad code.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:
```diff
@@ -164,4 +170,305 @@ trait SQLQueryTestHelper extends Logging {
     }
   }
+
+  /** A test case. */
```
I assume these are just code move around, no actual changes?

andylam-db replied:
Yes, that's correct.

cloud-fan commented:
> It's maybe a bit weird that a test in spark/connector uses/affects test files in spark/core.

I think it's OK. We have JDBC tests in `spark/connector` that depend on tests in `sql/core`

> There's some code duplication to read golden files, but I think not too much.

Is there a way to refactor it?

andylam-db committed:
Merge branch 'master' into crossdbms

andylam-db commented:
> Is there a way to refactor it?

I tried to move as much code as possible to the trait `SQLQueryHelper`. It's 20+ lines, I don't think it's an issue.

andylam-db commented:
@cloud-fan Can we merge?

cloud-fan commented:
thanks, merging to master!
